### ISSUE TITLE
DietPi-PREP: Failed to get canonical path of `/dev/simfs'.

### ISSUE THREAD

**Issue by LexiconCode** [MichaIng/DietPi#2841]
#### Details:
- Date           | Tue 21 May 16:02:04 BST 2019
- Bug report     | N/A
- DietPi version | v6.24.1 (MichaIng/master)
- Img creator    | n/a
- Pre-image      | n/a
- SBC device     | NULL (index=20)
- Kernel version | #1 SMP Wed Feb 27 09:04:24 MSK 2019
- Distro         | stretch (index=4)
- Command        | update-grub
- Exit code      | 1
- Software title | DietPi-PREP

#### Steps to reproduce:

1. Follow the instructions on [PREP_SYSTEM_FOR_DIETPI.sh | Automate](https://github.com/MichaIng/DietPi/issues/1285#issue-280771944)

#### Expected behaviour:

- The script should convert to a dietpi system.
- the PREP_SYSTEM_FOR_DIETPI.s expects '/dev/simf' file path
#### Actual behaviour:

- failed to get canonical path of `/dev/simfs'
- `/dev/simfs' file path does not actually exist. Therefore script fails.

#### Extra details:

- This is being installed on a commercial VPS (OpenVZ) 
- The original system was Debian 8 64 Bit which I updated in place to Debian 9 64 Bit using this [Guide](https://phoenixnap.com/kb/how-to-upgrade-debian-8-jessie-to-debian-9-stretch)
- Workaround `mkdir /dev/simfs` seems to allow the script to complete but the system becomes unbootable.

#### Additional logs:
```
Log file contents:
/usr/sbin/grub-probe: error: failed to get canonical path of `/dev/simfs'.
```

**Comment by MichaIng** [comment]
@LexiconCode 
Many thanks for your report.

Yeah an initial Jessie system with some custom VPS setup always includes unexpected results. `/dev/simfs` might be something like the virtual file system of the VPS boot partition (with bootloader on it => grub), although strange then that it does not exist. Possibly some package is required or something has gone with the Stretch upgrade.

Btw I recommend you use DietPi-PREP to upgrade Jessie => Stretch. It supports this! It detects Jessie and allows you to select Stretch and even Buster as target repo.

The benefit is that since DietPi-PREP vastly reduces the package amount, the distro upgrade is much faster and less risky. However if the manual upgrade went fine, the result will be the same.

Can you paste:
```
lsblk
df
```

**Comment by LexiconCode** [comment]
Further testing yields of the system is unbootable and I can't access via SSH to assess. I will try the `use DietPi-PREP to upgrade Jessie => Stretch.` with a quick reinstall.

**Comment by MichaIng** [comment]
@LexiconCode 
Since the error appeared during `update-grub` command which is the bootloader update, it is very expected that a failure there results in an unbootable system. Please inspect the drive and partition setup of the base image first!

Ah and also check:
`cat /etc/fstab`
- Possibly it contains specific entries/options required for the VPS to boot, which are not added by DietPi-Drive_Manager.

**Comment by LexiconCode** [comment]
lsblk
df

```
Filesystem     1K-blocks   Used Available Use% Mounted on
/dev/simfs      62914560 594996  62319564   1% /
devtmpfs         3145728      0   3145728   0% /dev
tmpfs            3145728      0   3145728   0% /dev/shm
tmpfs            3145728   9328   3136400   1% /run
tmpfs               5120      0      5120   0% /run/lock
tmpfs            3145728      0   3145728   0% /sys/fs/cgroup
tmpfs            3145728      0   3145728   0% /tmp
none             3145728      0   3145728   0% /run/shm
```

**Comment by LexiconCode** [comment]
cat /etc/fstab
```
proc  /proc       proc    defaults    0    0
none  /dev/pts    devpts  rw,gid=5,mode=620    0    0
none  /run/shm    tmpfs   defaults    0    0
```

**Comment by MichaIng** [comment]
@LexiconCode 
Strange, no /dev/simfs mount entry in fstab 🤔. There must be some other mechanism mounting it.

Which base image did you install on the VPS? Are there read-to-run Debian Jessie images available?

It needs to be checked which mechanism mounts the rootfs (/dev/simfs). Is grub even the bootloader?
`dpkg -l | grep grub`

And the file system type could be important as well: `df -T`
- Perhaps it can be added to fstab and allows grub boot, but needs a special entry.
______
Strange also: While the mount is active it seems to be expected that the path does not exist (any more): https://stackoverflow.com/questions/26326123/dev-simfs-no-such-file-or-directory-while-opening-filesystem

**Comment by LexiconCode** [comment]
Base Image: Debian 8 64 Bit
Provider: https://www.x4servers.com/

`dpkg -l | grep grub` returns nothing
df -T
```

Filesystem     Type     1K-blocks   Used Available Use% Mounted on
/dev/simfs     simfs     62914560 595080  62319480   1% /
devtmpfs       devtmpfs   3145728      0   3145728   0% /dev
tmpfs          tmpfs      3145728      0   3145728   0% /dev/shm
tmpfs          tmpfs      3145728   9328   3136400   1% /run
tmpfs          tmpfs         5120      0      5120   0% /run/lock
tmpfs          tmpfs      3145728      0   3145728   0% /sys/fs/cgroup
tmpfs          tmpfs      3145728      0   3145728   0% /tmp
none           tmpfs      3145728      0   3145728   0% /run/shm

```
I would be willing to give you SSH access.

**Comment by MichaIng** [comment]
@LexiconCode 
I think there is not very much we can do. This VPS seems to use a custom bootloader and custom virtual drive. We could hack this to working perhaps, but you will run into issues again, e.g. running DietPi drive manager and such.

However to play around:
- `ls -al /boot` # To perhaps get a hint about the boot loader
- In DietPi-PREP script you need to remove the `grub-pc`/`grub-efi` ATP install lines.
- Check `dpkg-query -s initramfs-tools`. If it is not installed, then remove the line containing `initramfs-tools` from DietPi-PREP script as well.
- After DietPi-PREP has finished, you need to remove the `/dev/simfs` line from `/etc/fstab`, since this does not exist and the rootfs is mounted differently.
_______
About SimFS in general: https://github.com/spcl/SimFS

**Comment by LexiconCode** [comment]
That's a shame but understandable. I'd rather find another VPS that works.

ls -al /boot
```
total 8
drwxr-xr-x  2 root root 4096 Nov 26  2016 .
drwxr-xr-x 23 root root 4096 May 21 11:24 ..
```
dpkg-query -s initramfs-tools.
`dpkg-query: package 'initramfs-tools.' is not installed and no information is available`

**Comment by LexiconCode** [comment]
I wonder if it's worthwhile Creating PREP_SYSTEM_FOR_DIETPI.sh for VPS? The DietPi-Drive_Manager isn't necessary for a VPS. Backups are handled by the VPS and they typically only have one virtual drive.

**Comment by MichaIng** [comment]
@LexiconCode
Okay all bootloader stuff seems to run completely outside the virtual drive.

You can give it a try with skipping grub and initramfs install in DietPi-PREP script

**Comment by LexiconCode** [comment]
Sure I'll give that a try. Verify I should be using the following
`wget https://raw.githubusercontent.com/MichaIng/DietPi/jessie-support/PREP_SYSTEM_FOR_DIETPI.sh -O PREP_SYSTEM_FOR_DIETPI.sh
`

**Comment by LexiconCode** [comment]
@MichaIng 
This may be because I didn't make the edits correctly. I commented out lines only. However when a script completes. VPS does not boot.
```
nano /etc/fstab
-bash: /usr/bin/nano: No such file or directory
```
https://gist.github.com/LexiconCode/bd28190e351fae3b8fc34a82dc06cb6b

**Comment by LexiconCode** [comment]
After further investigation I come across a strange scenario as fstab seems to exist(I made a backup as well) however when I try to edit it it doesn't exist.


ls /etc/fstab
/etc/fstab

 nano /etc/fstab
-bash: /usr/bin/nano: No such file or directory

**Comment by MichaIng** [comment]
@LexiconCode 
<s>`nano` is what does not exist, although strange because if it was not installed, the error message should look different. Commenting lines in an array might create issues not sure.

But to verify: `sudo apt install nano`

Ah no there is one syntax issue in your version:
This line: https://gist.github.com/LexiconCode/bd28190e351fae3b8fc34a82dc06cb6b#file-prep_system_for_dietpi-sh-L816
- Change `elif` to `if`</s>
______
Ah no better: Uncomment these two lines: https://gist.github.com/LexiconCode/bd28190e351fae3b8fc34a82dc06cb6b#file-prep_system_for_dietpi-sh-L797-L799
- The kernel should be still required.
______
And about the DietPi-PREP branch: https://raw.githubusercontent.com/MichaIng/DietPi/jessie-support/PREP_SYSTEM_FOR_DIETPI.sh
- You can still use the master version: https://raw.githubusercontent.com/MichaIng/DietPi/master/PREP_SYSTEM_FOR_DIETPI.sh
- Both are identical however. The jessie-support branch is the one that at lastest with v6.26 will support Jessie systems to RUN, but DietPi-PREP will still support using Jessie images as basis, but will force upgrading it to Stretch at least.

**Comment by LexiconCode** [comment]
I uncommented those two lines
https://gist.github.com/LexiconCode/bd28190e351fae3b8fc34a82dc06cb6b#file-prep_system_for_dietpi-sh-L797-L799

but receive
./PREP_SYSTEM_FOR_DIETPI.sh: line 1765: syntax error near unexpected token `}'
./PREP_SYSTEM_FOR_DIETPI.sh: line 1765: `       }'

**Comment by MichaIng** [comment]
@LexiconCode 
This line as well needs to be uncommented: https://gist.github.com/LexiconCode/bd28190e351fae3b8fc34a82dc06cb6b#file-prep_system_for_dietpi-sh-L789

**Comment by LexiconCode** [comment]
@MichaIng
I just want to say thanks for your help I really appreciate it! 
The script parses as expected with the changes that you recommended.

> After further investigation I come across a strange scenario as fstab seems to exist(I made a backup as well) however when I try to edit it it doesn't exist.

I'm now experiencing the issue as described in the above quote. I have not rebooted yet because I cannot check the following.

>     * After DietPi-PREP has finished, you need to remove the `/dev/simfs` line from `/etc/fstab`, since this does not exist and the rootfs is mounted differently.

Reference
https://unix.stackexchange.com/questions/441493/bash-usr-bin-nano-no-such-file-or-directory
May be an issue transitioning from Debian 8 to 9.

**Update** `hash -r` fixes the issue with nano.

nano /etc/fstab
```
# Please use "dietpi-drive_manager" to setup mounts
#----------------------------------------------------------------
# NETWORK
#----------------------------------------------------------------


#----------------------------------------------------------------
# TMPFS
#----------------------------------------------------------------
tmpfs /tmp tmpfs defaults,noatime,nodev,nosuid,mode=1777 0 0
tmpfs /var/log tmpfs defaults,size=50m,noatime,nodev,nosuid,mode=1777 0 0
tmpfs /DietPi tmpfs defaults,size=10m,noatime,nodev,nosuid,mode=1777 0 0

#----------------------------------------------------------------
# MISC: encryptfs, vboxsf (VirtualBox shared folder), bind mounts
#----------------------------------------------------------------


#----------------------------------------------------------------
# SWAPFILE
#----------------------------------------------------------------


#----------------------------------------------------------------
# PHYSICAL DRIVES
#----------------------------------------------------------------

```

**Comment by LexiconCode** [comment]
> 
> 
> @MichaIng
> I just want to say thanks for your help I really appreciate it!
> The script parses as expected with the changes that you recommended.
> 
> > After further investigation I come across a strange scenario as fstab seems to exist(I made a backup as well) however when I try to edit it it doesn't exist.
> 
> I'm now experiencing the issue as described in the above quote. I have not rebooted yet because I cannot check the following.
> 
> > ```
> > * After DietPi-PREP has finished, you need to remove the `/dev/simfs` line from `/etc/fstab`, since this does not exist and the rootfs is mounted differently.
> > ```
reference
https://unix.stackexchange.com/questions/441493/bash-usr-bin-nano-no-such-file-or-directory

**Comment by MichaIng** [comment]
@LexiconCode 
> hash -r fixes the issue with nano.

Good find. So the $PATH location content seems to be cached and `hash -r` = `forget all remembered locations`.

Okay the fstab looks good. Was not 100% sure if the `/dev/simfs` was added. Could you paste:
`dpkg -l`
So just to see tat really all bootloader, initramfs etc packages that affect boot steps before kernel do not exist.
__________
> I wonder if it's worthwhile Creating PREP_SYSTEM_FOR_DIETPI.sh for VPS?

The problem is that all VPS work differently. E.g. myVirtualServer where dietpi.com runs on works with default Debian images and default x86 DietPi (hence DietPi-PREP) out of the box. It loads the standard bootloader and initramfs that you need for real machines. So the only possibility is to create images or add individual install paths for all VPS separately, at least for those that definitely need customisation compared to default x86. DietPi-Drive_Manager still has use for swap file, fsck and some stuff, although in case of simfs that should not work 🤔 Network and tmpfs mounts can be at least definitely handled. The issue with fstab is just simple: Only add /dev/* sources that really exist, regardless of `df` output. But according to your paste this already seems to be the case.

**Comment by LexiconCode** [comment]
dpkg -l
```
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                                            Version                              Architecture                         Description
+++-===============================================================-====================================-====================================-===================================================================================================================================
ii  adduser                                                         3.115                                all                                  add and remove users and groups
ii  apt                                                             1.4.9                                amd64                                commandline package manager
ii  apt-transport-https                                             1.4.9                                amd64                                https download transport for APT
ii  apt-utils                                                       1.4.9                                amd64                                package management related utility programs
ii  base-files                                                      9.9+deb9u9                           amd64                                Debian base system miscellaneous files
ii  base-passwd                                                     3.5.43                               amd64                                Debian base system master password and group files
ii  bash                                                            4.4-5                                amd64                                GNU Bourne Again SHell
ii  bash-completion                                                 1:2.1-4.3                            all                                  programmable completion for the bash shell
ii  bc                                                              1.06.95-9+b3                         amd64                                GNU bc arbitrary precision calculator language
ii  bsdutils                                                        1:2.29.2-1+deb9u1                    amd64                                basic utilities from 4.4BSD-Lite
ii  bzip2                                                           1.0.6-8.1                            amd64                                high-quality block-sorting file compressor - utilities
ii  ca-certificates                                                 20161130+nmu1+deb9u1                 all                                  Common CA certificates
ii  console-setup                                                   1.164                                all                                  console font and keymap setup program
ii  console-setup-linux                                             1.164                                all                                  Linux specific part of console-setup
ii  coreutils                                                       8.26-3                               amd64                                GNU core utilities
ii  cpio                                                            2.11+dfsg-6                          amd64                                GNU cpio -- a program to manage archives of files
ii  cron                                                            3.0pl1-128+deb9u1                    amd64                                process scheduling daemon
ii  curl                                                            7.52.1-5+deb9u9                      amd64                                command line tool for transferring data with URL syntax
ii  dash                                                            0.5.8-2.4                            amd64                                POSIX-compliant shell
ii  debconf                                                         1.5.61                               all                                  Debian configuration management system
ii  debian-archive-keyring                                          2017.5                               all                                  GnuPG archive keys of the Debian archive
ii  debianutils                                                     4.8.1.1                              amd64                                Miscellaneous utilities specific to Debian
ii  diffutils                                                       1:3.5-3                              amd64                                File comparison utilities
ii  dirmngr                                                         2.1.18-8~deb9u4                      amd64                                GNU privacy guard - network certificate management service
ii  dmsetup                                                         2:1.02.137-2                         amd64                                Linux Kernel Device Mapper userspace library
ii  dpkg                                                            1.18.25                              amd64                                Debian package management system
ii  dropbear-bin                                                    2016.74-5+deb9u1                     amd64                                lightweight SSH2 server and client - command line tools
ii  dropbear-run                                                    2016.74-5+deb9u1                     all                                  lightweight SSH2 server and client - startup scripts
ii  e2fslibs:amd64                                                  1.43.4-2                             amd64                                ext2/ext3/ext4 file system libraries
ii  e2fsprogs                                                       1.43.4-2                             amd64                                ext2/ext3/ext4 file system utilities
ii  ethtool                                                         1:4.8-1+b1                           amd64                                display or change Ethernet device settings
ii  fake-hwclock                                                    0.11                                 all                                  Save/restore system clock on machines without working RTC hardware
ii  findutils                                                       4.6.0+git+20161106-2                 amd64                                utilities for finding files--find, xargs
ii  gawk                                                            1:4.1.4+dfsg-1                       amd64                                GNU awk, a pattern scanning and processing language
ii  gcc-4.8-base:amd64                                              4.8.4-1                              amd64                                GCC, the GNU Compiler Collection (base package)
ii  gcc-4.9-base:amd64                                              4.9.2-10                             amd64                                GCC, the GNU Compiler Collection (base package)
ii  gcc-6-base:amd64                                                6.3.0-18+deb9u1                      amd64                                GCC, the GNU Compiler Collection (base package)
ii  gnupg                                                           2.1.18-8~deb9u4                      amd64                                GNU privacy guard - a free PGP replacement
ii  gnupg-agent                                                     2.1.18-8~deb9u4                      amd64                                GNU privacy guard - cryptographic agent
ii  gpgv                                                            2.1.18-8~deb9u4                      amd64                                GNU privacy guard - signature verification tool
ii  grep                                                            2.27-2                               amd64                                GNU grep, egrep and fgrep
ii  gzip                                                            1.6-5+b1                             amd64                                GNU compression utilities
ii  hostname                                                        3.18+b1                              amd64                                utility to set/show the host name or domain name
ii  htop                                                            2.0.2-1                              amd64                                interactive processes viewer
ii  ifupdown                                                        0.8.19                               amd64                                high level tools to configure network interfaces
ii  init-system-helpers                                             1.48                                 all                                  helper tools for all init systems
ii  initramfs-tools                                                 0.130                                all                                  generic modular initramfs generator (automation)
ii  initramfs-tools-core                                            0.130                                all                                  generic modular initramfs generator (core tools)
ii  initscripts                                                     2.88dsf-59.9                         amd64                                scripts for initializing and shutting down the system
ii  insserv                                                         1.14.0-5.4+b1                        amd64                                boot sequence organizer using LSB init.d script dependency information
ii  install-info                                                    6.3.0.dfsg.1-1+b2                    amd64                                Manage installed documentation in info format
ii  iproute2                                                        4.9.0-1+deb9u1                       amd64                                networking and traffic control tools
ii  iputils-ping                                                    3:20161105-1                         amd64                                Tools to test the reachability of network hosts
ii  isc-dhcp-client                                                 4.3.5-3+deb9u1                       amd64                                DHCP client for automatically obtaining an IP address
ii  kbd                                                             2.0.3-2+b1                           amd64                                Linux console font and keytable utilities
ii  keyboard-configuration                                          1.164                                all                                  system-wide keyboard preferences
ii  klibc-utils                                                     2.0.4-9                              amd64                                small utilities built with klibc for early boot
ii  kmod                                                            23-2                                 amd64                                tools for managing Linux kernel modules
ii  libacl1:amd64                                                   2.2.52-3+b1                          amd64                                Access control list shared library
ii  libapparmor1:amd64                                              2.11.0-3+deb9u2                      amd64                                changehat AppArmor library
ii  libapt-inst2.0:amd64                                            1.4.9                                amd64                                deb package format runtime library
ii  libapt-pkg5.0:amd64                                             1.4.9                                amd64                                package management runtime library
ii  libassuan0:amd64                                                2.4.3-2                              amd64                                IPC library for the GnuPG components
ii  libattr1:amd64                                                  1:2.4.47-2+b2                        amd64                                Extended attribute shared library
ii  libaudit-common                                                 1:2.6.7-2                            all                                  Dynamic library for security auditing - common files
ii  libaudit1:amd64                                                 1:2.6.7-2                            amd64                                Dynamic library for security auditing
ii  libblkid1:amd64                                                 2.29.2-1+deb9u1                      amd64                                block device ID library
ii  libbz2-1.0:amd64                                                1.0.6-8.1                            amd64                                high-quality block-sorting file compressor library - runtime
ii  libc-bin                                                        2.24-11+deb9u4                       amd64                                GNU C Library: Binaries
ii  libc-l10n                                                       2.24-11+deb9u4                       all                                  GNU C Library: localization files
ii  libc6:amd64                                                     2.24-11+deb9u4                       amd64                                GNU C Library: Shared libraries
ii  libcap-ng0:amd64                                                0.7.7-3+b1                           amd64                                An alternate POSIX capabilities library
ii  libcap2:amd64                                                   1:2.25-1                             amd64                                POSIX 1003.1e capabilities (library)
ii  libcomerr2:amd64                                                1.43.4-2                             amd64                                common error description library
ii  libcryptsetup4:amd64                                            2:1.7.3-4                            amd64                                disk encryption support - shared library
ii  libcurl3:amd64                                                  7.52.1-5+deb9u9                      amd64                                easy-to-use client-side URL transfer library (OpenSSL flavour)
ii  libcurl3-gnutls:amd64                                           7.52.1-5+deb9u9                      amd64                                easy-to-use client-side URL transfer library (GnuTLS flavour)
ii  libdb5.3:amd64                                                  5.3.28-12+deb9u1                     amd64                                Berkeley v5.3 Database Libraries [runtime]
ii  libdebconfclient0:amd64                                         0.227                                amd64                                Debian Configuration Management System (C-implementation library)
ii  libdevmapper1.02.1:amd64                                        2:1.02.137-2                         amd64                                Linux Kernel Device Mapper userspace library
ii  libdns-export162                                                1:9.10.3.dfsg.P4-12.3+deb9u5         amd64                                Exported DNS Shared Library
ii  libelf1:amd64                                                   0.168-1                              amd64                                library to read and write ELF files
ii  libfdisk1:amd64                                                 2.29.2-1+deb9u1                      amd64                                fdisk partitioning library
ii  libffi6:amd64                                                   3.2.1-6                              amd64                                Foreign Function Interface library runtime
ii  libgcc1:amd64                                                   1:6.3.0-18+deb9u1                    amd64                                GCC support library
ii  libgcrypt20:amd64                                               1.7.6-2+deb9u3                       amd64                                LGPL Crypto library - runtime library
ii  libgmp10:amd64                                                  2:6.1.2+dfsg-1                       amd64                                Multiprecision arithmetic library
ii  libgnutls30:amd64                                               3.5.8-5+deb9u4                       amd64                                GNU TLS library - main runtime library
ii  libgpg-error0:amd64                                             1.26-2                               amd64                                library for common error values and messages in GnuPG components
ii  libgssapi-krb5-2:amd64                                          1.15-1+deb9u1                        amd64                                MIT Kerberos runtime libraries - krb5 GSS-API Mechanism
ii  libhogweed4:amd64                                               3.3-1+b2                             amd64                                low level cryptographic library (public-key cryptos)
ii  libidn11:amd64                                                  1.33-1                               amd64                                GNU Libidn library, implementation of IETF IDN specifications
ii  libidn2-0:amd64                                                 0.16-1+deb9u1                        amd64                                Internationalized domain names (IDNA2008) library
ii  libip4tc0:amd64                                                 1.6.0+snapshot20161117-6             amd64                                netfilter libip4tc library
ii  libisc-export160                                                1:9.10.3.dfsg.P4-12.3+deb9u5         amd64                                Exported ISC Shared Library
ii  libk5crypto3:amd64                                              1.15-1+deb9u1                        amd64                                MIT Kerberos runtime libraries - Crypto Library
ii  libkeyutils1:amd64                                              1.5.9-9                              amd64                                Linux Key Management Utilities (library)
ii  libklibc                                                        2.0.4-9                              amd64                                minimal libc subset for use with initramfs
ii  libkmod2:amd64                                                  23-2                                 amd64                                libkmod shared library
ii  libkrb5-3:amd64                                                 1.15-1+deb9u1                        amd64                                MIT Kerberos runtime libraries
ii  libkrb5support0:amd64                                           1.15-1+deb9u1                        amd64                                MIT Kerberos runtime libraries - Support library
ii  libksba8:amd64                                                  1.3.5-2                              amd64                                X.509 and CMS support library
ii  libldap-2.4-2:amd64                                             2.4.44+dfsg-5+deb9u2                 amd64                                OpenLDAP libraries
ii  libldap-common                                                  2.4.44+dfsg-5+deb9u2                 all                                  OpenLDAP common files for libraries
ii  liblocale-gettext-perl                                          1.07-3+b1                            amd64                                module using libc functions for internationalization in Perl
ii  liblz4-1:amd64                                                  0.0~r131-2+b1                        amd64                                Fast LZ compression algorithm library - runtime
ii  liblzma5:amd64                                                  5.2.2-1.2+b1                         amd64                                XZ-format compression library
ii  libmnl0:amd64                                                   1.0.4-2                              amd64                                minimalistic Netlink communication library
ii  libmount1:amd64                                                 2.29.2-1+deb9u1                      amd64                                device mounting library
ii  libmpfr4:amd64                                                  3.1.5-1                              amd64                                multiple precision floating-point computation
ii  libncurses5:amd64                                               6.0+20161126-1+deb9u2                amd64                                shared libraries for terminal handling
ii  libncursesw5:amd64                                              6.0+20161126-1+deb9u2                amd64                                shared libraries for terminal handling (wide character support)
ii  libnettle6:amd64                                                3.3-1+b2                             amd64                                low level cryptographic library (symmetric and one-way cryptos)
ii  libnewt0.52:amd64                                               0.52.19-1+b1                         amd64                                Not Erik's Windowing Toolkit - text mode windowing with slang
ii  libnghttp2-14:amd64                                             1.18.1-1                             amd64                                library implementing HTTP/2 protocol (shared library)
ii  libnpth0:amd64                                                  1.3-1                                amd64                                replacement for GNU Pth using system threads
ii  libp11-kit0:amd64                                               0.23.3-2                             amd64                                library for loading and coordinating access to PKCS#11 modules - runtime
ii  libpam-modules:amd64                                            1.1.8-3.6                            amd64                                Pluggable Authentication Modules for PAM
ii  libpam-modules-bin                                              1.1.8-3.6                            amd64                                Pluggable Authentication Modules for PAM - helper binaries
ii  libpam-runtime                                                  1.1.8-3.6                            all                                  Runtime support for the PAM library
ii  libpam0g:amd64                                                  1.1.8-3.6                            amd64                                Pluggable Authentication Modules library
ii  libparted2:amd64                                                3.2-17                               amd64                                disk partition manipulator - shared library
ii  libpcre3:amd64                                                  2:8.39-3                             amd64                                Old Perl 5 Compatible Regular Expression Library - runtime files
ii  libpopt0:amd64                                                  1.16-10+b2                           amd64                                lib for parsing cmdline parameters
ii  libprocps6:amd64                                                2:3.3.12-3+deb9u1                    amd64                                library for accessing process information from /proc
ii  libpsl5:amd64                                                   0.17.0-3                             amd64                                Library for Public Suffix List (shared libraries)
ii  libreadline7:amd64                                              7.0-3                                amd64                                GNU readline and history libraries, run-time libraries
ii  librtmp1:amd64                                                  2.4+20151223.gitfa8646d.1-1+b1       amd64                                toolkit for RTMP streams (shared library)
ii  libsasl2-2:amd64                                                2.1.27~101-g0780600+dfsg-3           amd64                                Cyrus SASL - authentication abstraction library
ii  libsasl2-modules-db:amd64                                       2.1.27~101-g0780600+dfsg-3           amd64                                Cyrus SASL - pluggable authentication modules (DB)
ii  libseccomp2:amd64                                               2.3.1-2.1+deb9u1                     amd64                                high level interface to Linux seccomp filter
ii  libselinux1:amd64                                               2.6-3+b3                             amd64                                SELinux runtime shared libraries
ii  libsemanage-common                                              2.6-2                                all                                  Common files for SELinux policy management libraries
ii  libsemanage1:amd64                                              2.6-2                                amd64                                SELinux policy management library
ii  libsepol1:amd64                                                 2.6-2                                amd64                                SELinux library for manipulating binary security policies
ii  libsigsegv2:amd64                                               2.10-5                               amd64                                Library for handling page faults in a portable way
ii  libslang2:amd64                                                 2.3.1-5                              amd64                                S-Lang programming library - runtime version
ii  libsmartcols1:amd64                                             2.29.2-1+deb9u1                      amd64                                smart column output alignment library
ii  libsqlite3-0:amd64                                              3.16.2-5+deb9u1                      amd64                                SQLite 3 shared library
ii  libss2:amd64                                                    1.43.4-2                             amd64                                command-line interface parsing library
ii  libssh2-1:amd64                                                 1.7.0-1+deb9u1                       amd64                                SSH2 client-side library
ii  libssl1.0.2:amd64                                               1.0.2r-1~deb9u1                      amd64                                Secure Sockets Layer toolkit - shared libraries
ii  libssl1.1:amd64                                                 1.1.0j-1~deb9u1                      amd64                                Secure Sockets Layer toolkit - shared libraries
ii  libstdc++6:amd64                                                6.3.0-18+deb9u1                      amd64                                GNU Standard C++ Library v3
ii  libsystemd0:amd64                                               232-25+deb9u11                       amd64                                systemd utility library
ii  libtasn1-6:amd64                                                4.10-1.1+deb9u1                      amd64                                Manage ASN.1 structures (runtime)
ii  libtinfo5:amd64                                                 6.0+20161126-1+deb9u2                amd64                                shared low-level terminfo library for terminal handling
ii  libudev1:amd64                                                  232-25+deb9u11                       amd64                                libudev shared library
ii  libunistring0:amd64                                             0.9.6+really0.9.3-0.1                amd64                                Unicode string library for C
ii  libusb-1.0-0:amd64                                              2:1.0.21-1                           amd64                                userspace USB programming library
ii  libustr-1.0-1:amd64                                             1.0.4-6                              amd64                                Micro string library: shared library
ii  libuuid1:amd64                                                  2.29.2-1+deb9u1                      amd64                                Universally Unique ID library
ii  linux-base                                                      4.5                                  all                                  Linux image base package
ii  linux-image-4.9.0-9-amd64                                       4.9.168-1+deb9u2                     amd64                                Linux 4.9 for 64-bit PCs
ii  linux-image-amd64                                               4.9+80+deb9u7                        amd64                                Linux for 64-bit PCs (meta-package)
ii  locales                                                         2.24-11+deb9u4                       all                                  GNU C Library: National Language (locale) data [support]
ii  login                                                           1:4.4-4.1                            amd64                                system login tools
ii  lsb-base                                                        9.20161125                           all                                  Linux Standard Base init script functionality
rc  module-init-tools                                               18-3                                 all                                  transitional dummy package (module-init-tools to kmod)
ii  mount                                                           2.29.2-1+deb9u1                      amd64                                tools for mounting and manipulating filesystems
ii  multiarch-support                                               2.24-11+deb9u4                       amd64                                Transitional package to ensure multiarch compatibility
ii  nano                                                            2.7.4-1                              amd64                                small, friendly text editor inspired by Pico
ii  ncurses-base                                                    6.0+20161126-1+deb9u2                all                                  basic terminal type definitions
ii  ncurses-bin                                                     6.0+20161126-1+deb9u2                amd64                                terminal-related programs and man pages
ii  openssl                                                         1.1.0j-1~deb9u1                      amd64                                Secure Sockets Layer toolkit - cryptographic utility
ii  p7zip                                                           16.02+dfsg-3+deb9u1                  amd64                                7zr file archiver with high compression ratio
ii  parted                                                          3.2-17                               amd64                                disk partition manipulator
ii  passwd                                                          1:4.4-4.1                            amd64                                change and administer password and group data
ii  perl-base                                                       5.24.1-3+deb9u5                      amd64                                minimal Perl system
ii  pinentry-curses                                                 1.0.0-2                              amd64                                curses-based PIN or pass-phrase entry dialog for GnuPG
ii  procps                                                          2:3.3.12-3+deb9u1                    amd64                                /proc file system utilities
ii  psmisc                                                          22.21-2.1+b2                         amd64                                utilities that use the proc file system
ii  readline-common                                                 7.0-3                                all                                  GNU readline and history libraries, common files
ii  resolvconf                                                      1.79                                 all                                  name server information handler
ii  sed                                                             4.4-1                                amd64                                GNU stream editor for filtering/transforming text
ii  sensible-utils                                                  0.0.9+deb9u1                         all                                  Utilities for sensible alternative selection
ii  startpar                                                        0.59-3.1                             amd64                                run processes in parallel and multiplex their output
ii  sudo                                                            1.8.19p1-2.1                         amd64                                Provide limited super user privileges to specific users
ii  systemd                                                         232-25+deb9u11                       amd64                                system and service manager
ii  systemd-sysv                                                    232-25+deb9u11                       amd64                                system and service manager - SysV links
ii  sysv-rc                                                         2.88dsf-59.9                         all                                  System-V-like runlevel change mechanism
ii  sysvinit-utils                                                  2.88dsf-59.9                         amd64                                System-V-like utilities
ii  tar                                                             1.29b-1.1                            amd64                                GNU version of the tar archiving utility
ii  tzdata                                                          2019a-0+deb9u1                       all                                  time zone and daylight-saving time data
ii  udev                                                            232-25+deb9u11                       amd64                                /dev/ and hotplug management daemon
ii  unzip                                                           6.0-21+deb9u1                        amd64                                De-archiver for .zip files
ii  usbutils                                                        1:007-4+b1                           amd64                                Linux USB utilities
ii  util-linux                                                      2.29.2-1+deb9u1                      amd64                                miscellaneous system utilities
ii  wget                                                            1.18-5+deb9u3                        amd64                                retrieves files from the web
ii  whiptail                                                        0.52.19-1+b1                         amd64                                Displays user-friendly dialog boxes from shell scripts
ii  xkb-data                                                        2.19-1+deb9u1                        all                                  X Keyboard Extension (XKB) configuration data
ii  zlib1g:amd64                                                    1:1.2.8.dfsg-5                       amd64                                compression library - runtime

```

I've tried rebooting and it doesn't boot :/

Interesting. if I don't reboot and logout of the SSH session then log back in the passwords accepted but then SSH session immediately ends without an error message.

**Comment by MichaIng** [comment]
> ii  initramfs-tools
> ii  initramfs-tools-core 

Not sure why but they're still there. Need to be purged: `apt purge initramfs-tools initramfs-tools-core`

And two other leftovers: `apt purge module-init-tools insserv`

Then: `apt autoremove --purge`

Still no guarantee that it works, but this is all I can think of.

> Interesting. if I don't reboot and logout of the SSH session then log back in the passwords accepted but then SSH session immediately ends without an error message.

Jep possibly the login script fails due to missing RAMdisk... Hmm, actually only the login script should exit, not the whole session. However a reboot is required anyway.
_______
That change allows to exit and login SSH before reboot: https://github.com/MichaIng/DietPi/commit/4acae8ab95f1b8f6f1e2ce32e4891caf78a94d75
- That was actually a good find. Not desired that without RAMdisk login is not possible at all 👍.
- Related changelog: https://github.com/MichaIng/DietPi/commit/d56f4801203d9ffc3520cd29c15e9dd7172c777d

**Comment by LexiconCode** [comment]
No boot, but a valiant effort!

Well it was a good try and I'm glad we discovered a bug. I guess I'll try to pursue getting my money back.

**Comment by MichaIng** [comment]
@LexiconCode 
> No boot, but a valiant effort!
Well it was a good try and I'm glad we discovered a bug.

Jep worth it, also good to learn a bid about different VPS instances.

> I guess I'll try to pursue getting my money back.

You could also ask them to help with/provide info info about special needs to boot their machines. Possibly some special required packages or files that need to stay or any that definitely need to be absent, like bootloader and most likely initramfs in this case.

**Comment by MichaIng** [comment]
Is it this one? https://openvz.org/
Although you link: https://www.x4servers.com/vps-hosting/
And they state `Platform Independent`. The empty `/boot` directory was on a fresh Jessie image they provide, right? And also `ls -l /` does not list `initrd.img` and/or `vmlinuz`?

**Comment by LexiconCode** [comment]
`boot` is empty.

`ls -l /` which does not list initrd.img and/or vmlinuz.
```
drwxr-xr-x  2 root root 4096 Nov 26  2016 bin
drwxr-xr-x  2 root root 4096 Nov 26  2016 boot
drwxr-xr-x  6 root  500  660 May 21 18:56 dev
drwxr-xr-x 85 root root 4096 May 21 18:56 etc
drwxr-xr-x  2 root root 4096 Nov 26  2016 home
drwxr-xr-x 13 root root 4096 May 21 18:56 lib
drwxr-xr-x  2 root root 4096 Nov 26  2016 lib64
drwx------  2 root root 4096 Nov 26  2016 lost+found
drwxr-xr-x  2 root root 4096 Nov 26  2016 media
drwxr-xr-x  2 root root 4096 Nov 26  2016 mnt
drwxr-xr-x  2 root root 4096 Nov 26  2016 opt
dr-xr-xr-x 58 root root    0 May 21 18:56 proc
drwx------  2 root root 4096 Nov 26  2016 root
drwxr-xr-x 17 root root  480 May 21 18:56 run
drwxr-xr-x  2 root root 4096 Nov 26  2016 sbin
drwxr-xr-x  2 root root 4096 Nov 26  2016 srv
drwxr-xr-x  7 root root    0 May 21 18:56 sys
drwxrwxrwt  7 root root  140 May 21 19:50 tmp
drwxr-xr-x 11 root root 4096 Nov 26  2016 usr
drwxr-xr-x 12 root root 4096 Nov 26  2016 var
```
Their control panel shows `OpenVZ` beyond that and I'm not sure. However contact with support indicates that they use solsuvm's official template to deploy their servers.

**Comment by MichaIng** [comment]
@LexiconCode 
Hmm SolusVM offers VMs on different basis, but they do not seem to support Debian: https://documentation.solusvm.com/display/DOCS/SolusVM+Installer++-+Version+4

Yeah however, as said the support might help when you ask them how to install/setup a custom Debian-based image.

**Comment by MichaIng** [comment]
I mark this issue as closed. Generic VPS support is for my knowledge a difficult thing, since every VPS invokes the image in a different way:
- Some may emulte that hardware only (like VirtualBox/VMware and such) and run fine with every arch-fitting Linux distro.
- Others emulate the bootloader too and invoke an external boot partition or initramfs for this, so any OS-internal kernel/firmware update can break things, if this does not fit to external bootloader, and especially any image-internal boot partition and related fstab entry might conflict.
- Again others might emulte the kernel (load an external one), like `qemu` does for ARM devices. Then the possible incompatibilities raise when the image comes with another kernel and software find and/or loads those modules/headers/... that do not fit the emulated kernel.

In theory we could add some env vars/menu that allows to skip initramfs, bootloader and/or kernel package installs, but if this then requires some other packages, flag files or special configs is again unknown, so currently out of scope IMO.
